src/exodus/sigmawallet.cpp

```diff
@@ -91,195 +72,193 @@ uint32_t SigmaWallet::CreateNextSeed(CKeyID &seedId, uint512& seed)
-        return false;
+    auto lastSlash = path.find_last_of('/');
+    if (lastSlash == std::string::npos) {
+        throw std::runtime_error("Fail to match BIP44 path");
     }
 
-    auto childData = path.substr(startPos);
-
-    auto isHardened = childData.back() == '\'';
-    if (isHardened) {
-        childData.resize(childData.size() - 1);
+    auto child = std::stol(path.substr(lastSlash + 1));
+    if (child > std::numeric_limits<uint32_t>::max()) {
+        throw std::runtime_error("Address index is exceed limit");
     }
 
-    if (childData.empty()) {
-        return false;
-    }
+    return child;
+}
 
-    auto childPos = std::stol(childData);
+secp_primitives::Scalar GetSerialFromPublicKey(
+    secp256k1_context const *context,
+    secp256k1_pubkey *pubkey)
+{
+    std::array<uint8_t, 32> pubkey_hash;
 
-    if (childPos > std::numeric_limits<uint32_t>::max()) {
-        return false;
+    static const unsigned char one[32] = {
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01
+    };
+
+    if (!secp256k1_ecdh(context, pubkey_hash.data(), pubkey, &one[0])) {
+        throw std::runtime_error("Unable to compute public key hash with secp256k1_ecdh.");
     }
 
-    child = {static_cast<uint32_t>(childPos), isHardened};
+    std::string zpts(ZEROCOIN_PUBLICKEY_TO_SERIALNUMBER);
+    std::array<uint8_t, sizeof(ZEROCOIN_PUBLICKEY_TO_SERIALNUMBER) - 1 +
+        std::tuple_size<decltype(pubkey_hash)>::value> pre;
+
+    auto ptr = std::copy(
+        reinterpret_cast<unsigned char const*>(zpts.data()),
+        reinterpret_cast<unsigned char const*>(zpts.data() + zpts.size()),
+        pre.data()
+    );
+
+    std::copy(pubkey_hash.begin(), pubkey_hash.end(), ptr);
+
+    std::array<unsigned char, CSHA256::OUTPUT_SIZE>  hash;
+    CSHA256().Write(pre.data(), pre.size()).Finalize(hash.data());
+
+    return Scalar(hash.data());
+}
 
-    return true;
 }
 
 uint32_t SigmaWallet::GetSeedIndex(CKeyID const &seedId)
 {
-    if (!pwalletMain->mapKeyMetadata.count(seedId)) {
+    LOCK(pwalletMain->cs_wallet);
+    auto it = pwalletMain->mapKeyMetadata.find(seedId);
+    if (it == pwalletMain->mapKeyMetadata.end()) {
         throw std::runtime_error("key not found");
     }
 
-    auto const &meta = pwalletMain->mapKeyMetadata[seedId];
-
     // parse last index
-    std::pair<uint32_t, bool> child;
-    if (!GetChildFromPath(meta.hdKeypath, child)) {
-        throw std::runtime_error("fail to parse HD key path");
-    }
-
-    if (child.second) {
-        throw std::runtime_error("hardened is not allowed");
+    uint32_t addressIndex;
+    try {
+        addressIndex = GetBIP44AddressIndex(it->second.hdKeypath);
+    } catch (std::runtime_error const &e) {
+        error("%s : fail to get child from, %s\n", __func__, e.what());
```

Comment:
We should use `PrintToLog` instead due to `error` used to generate error code for Bitcoin but we don't that functionality here.